cpp/ql/src/Likely Bugs/Format/NonConstantFormat.ql

```diff
@@ -17,6 +17,11 @@
 
 import semmle.code.cpp.ir.dataflow.TaintTracking
 import semmle.code.cpp.commons.Printf
+import semmle.code.cpp.security.FlowSources
+
+class UncalledFunction extends Function {
+  UncalledFunction() { not exists(Call c | c.getTarget() = this) }
```

Comment:
I think this should also account for functions that have function pointers taken, since they are presumably called later.  We can do this via:
```
and not exists(FunctionAccess fa | fa.getTarget() = this)
```
We might want to think about member function overrides that may never be called *directly* as well.

Reply:
My understanding (and I've observed this through prior experiments) that dataflow will know a function pointer is resolved, and know its target, but Call doesn't know a target in these cases, even if dataflow statically determine the target. This has royally bitten me in the past, but I'm not sure what the solution is or if it's on github's radar. 

If I understand your correctly, you are saying `FunctionAccess` resolves this issue though?

Reply:
Yes - `Function`, `Call` and `FunctionAccess` are all AST classes that really do little more than describe the code, they don't draw anything from dataflow.  So we either need to add the `FunctionAccess` case, or get information about whether the function is called from dataflow instead if it's possible to do that (I'm not sure this is possible and there's might be different issues with that approach).

For class method overrides, I think we need:
```
and not this.(MemberFunction).overrides(_)
```

Reply:
@MathiasVP, What's possible here? I generally have needed (do need) call targets to resolve when known statically for function pointers. Can we add an issue separately to fix that too? If so, I can move forward grabbing the call target (or whatever) and the query will be agnostic to any future update.

Reply:
It's not possible generally to know which calls can be resolved before actually starting dataflow since dataflow makes use of call contexts to resolve some calls. For example, the call target of `f` in this example isn't known before we start computing dataflow and know that we're coming from `g` which provides the function that's passed in as a function pointer:
```cpp
void g() {
  call(myFunction);
}

void call(void (*f)()) {
  f();
}
```

What the `FunctionAccess` solution does is to rule out any function whose pointer is passed around, and thus there's a very _very_ small chance that this function is ever called. Because:
- If it was called directly we'd have a `FunctionCall fc` such that `fc.getTarget()` was the function, and
- If it was called indirectly (i.e., via a function pointer) then someone must have, at some point in the program, passed the function pointer around prior to the call. i.e., there must be a `FunctionAccess fa` such that `fc.getTarget()` was the function.